File: 0015_3Sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        for (int i = 0; i < nums.size(); i++) {
            int left = i + 1;
            int right = nums.size() - 1;
            while(left < right) {
                if (nums[i] + nums[left] + nums[right] > 0) 
                    right--;
                else {
                    if (nums[i] + nums[left] + nums[right] < 0) 
                        left++;
                    else {
                        vector <int> triplet = {nums[i], nums[left], nums[right]};
                        result.push_back(triplet);
                        
                        while (left < right && nums[left] == triplet[1]) 
                            left++;
                        
                        while (left < right && nums[right] == triplet[2]) 
                            right--;
                        
                    }
                }
            }
            while (i + 1 < nums.size() && nums[i + 1] == nums[i]) 
                i++;
        }
        return result;
    }
};

File: 0015_3Sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        i = 0
        while i < len(nums):
            left = i + 1
            right = len(nums) - 1
            while left < right:
                if nums[i] + nums[left] + nums[right] > 0:
                    right += -1
                else:
                    if nums[i] + nums[left] + nums[right] < 0:
                        left += 1 
                    else:
                        triplet = [nums[i], nums[left], nums[right]]
                        result.append(triplet)
                        
                        while left < right and nums[left] == triplet[1]:
                            left += 1
                        while left < right and nums[right] == triplet[2]:
                            right += -1
            
            while i + 1 < len(nums) and nums[i + 1] == nums[i]:
                i += 1
            i += 1
        
        return result